graphql_compiler/interpreter/typedefs.py

```diff
@@ -0,0 +1,208 @@
+      of the database entry using which the property values can be looked up, or
+    - an instance of a custom class which has *some* of the values of the vertex properties, and
+      has sufficient information to look up the rest of them if they are ever requested.
+
+    The best choice of DataToken type is dependent on the specific use case, e.g. whether the data
+    is already available in Python memory, or is on a local disk, or is a network hop away.
+
+    Implementers are free to choose any DataToken type and the interpreter code will happily use it.
+    However, for the sake of easier debugging and testing using the built-in functionality in
+    this library, it is desirable to make DataToken be a deep-copyable type that implements
+    equality beyond a simple referential equality check.
+
+    ## The InterpreterAdapter API
+
+    The methods in the InterpreterAdapter API are all designed to support generator-style operation,
+    where data is produced and consumed only when required. Here is a high-level description of
+    the methods in the InterpreterAdapter API:
+    - get_tokens_of_type() produces an iterable of DataTokens of the type specified by its argument.
+      The calling function will wrap the DataTokens into a bookkeeping object called a DataContext,
+      where a particular token is currently active and specified in the "current_token" attribute.
+    - For an iterable of such DataContexts, project_property() can be used to get the value
+      of one of the properties on the vertex type represented by the currently active DataToken
+      in each DataContext; project_property() therefore returns an iterable of
+      tuples (data_context, value).
+    - project_neighbors() is similar: for an iterable of DataContexts and a specific edge name,
+      it returns an iterable (data_context, iterable_of_neighbor_tokens) where each result
+      contains the DataTokens of the neighboring vertices along that edge for the vertex whose
+      DataToken is currently active in that DataContext.
+    - can_coerce_to_type() is used to check whether a DataToken corresponding to one vertex type
+      can be safely converted into one representing a different vertex type. Given an iterable of
+      DataContexts and the name of the type to which the conversion is attempted, it produces
+      an iterable of tuples (data_context, can_coerce), where can_coerce is a boolean.
+
+    ## Performance and optimization opportunities
+
+    The design of the API, including its generator-style operation, enables a variety of
+    optimizations to either happen automatically or be available with minimal additional work.
+    A few simple examples:
+    - Interpreters perform lazy evaluation by default: if exactly 3 query results are requested,
+      then only the minimal data necessary to produce *exactly 3 results' worth* of outputs is loaded.
```

Comment:
```suggestion
    - Interpreters perform lazy evaluation by default: if 3 query results are requested,
      only the minimal data necessary to produce *exactly 3* results worth of outputs is loaded.
```
I think this makes the emphasis a bit more effective. Not positive about the apostrophe, but I don't think it is needed?

Reply:
:+1: to everything, including dropping the apostrophe.